### ISSUE TITLE
Keyboard shortcuts under Windows missing modifiers

### ISSUE THREAD

**Issue by Clorix1** [darktable-org/darktable#10729]
I've just updated from darktable v3.6.1 to v3.8.0 on Windows 10 and saw the note about needing to set up my custom keyboard shortcuts again due to the new shortcut system. However, I've not been able to successfully set any shortcuts.

The issues seems to be that it removes one of my modifier keys for some reason. For example, I want to set a shortcut for the **utility modules > import > add to library** command set to **Ctrl + Shift + Alt + I**, but the shortcuts preferences window shows **Shift + Ctrl + I** instead.

I thought maybe turning on the "enable fallbacks" option might make a difference, but it doesn't seem to do anything. I've also made sure to unplug all other controllers now that they and other devices can be used as shortcuts, but that also didn't work. I didn't unplug my Stream Deck as that's my primary interface when using darktable, but even pressing a shortcut key on the Stream Deck doesn't make it work.

I'm sure I must be missing something obvious about the new shortcut system. Or is this a bug with the new version?

**Comment by elstoc** [comment]
Seems fine on Linux so maybe this is a Windows issue or something specific to your setup. What happens if you do Alt+I, for example? Or does it remove a different modifier key each time?

**Comment by dterrahe** [comment]
Do triple-modifier key combinations (with all three of Shift, Ctrl _and_ Alt) generally work with your keyboard in other applications (or even in dt 3.6)? The new code mostly uses the standard Gtk toolkit method of detecting which modifiers are pressed, so if it works elsewhere it is surprising that it doesn't work here. But some keyboards physically can't detect more than 2 or 3 keypresses at the same time; a fourth one can get ghosted by the others. I'm not sure how common it is that modifier keys would suffer this issue though.

For me it works fine under Windows for modifiers, but my laptop keyboard has more ghosting issues than my usb gaming keyboard.

**Comment by Clorix1** [comment]
> Seems fine on Linux so maybe this is a Windows issue or something specific to your setup. What happens if you do Alt+I, for example? Or does it remove a different modifier key each time?

It seems random which modifier it removes.  I can't seem to find a pattern to it if there is any, except that it's always a single modifier that gets removed each time.  So in your example, Alt+I will simply become just "I".


> Do triple-modifier key combinations (with all three of Shift, Ctrl _and_ Alt) generally work with your keyboard in other applications (or even in dt 3.6)? The new code mostly uses the standard Gtk toolkit method of detecting which modifiers are pressed, so if it works elsewhere it is surprising that it doesn't work here. But some keyboards physically can't detect more than 2 or 3 keypresses at the same time; a fourth one can get ghosted by the others. I'm not sure how common it is that modifier keys would suffer this issue though.
> 
> For me it works fine under Windows for modifiers, but my laptop keyboard has more ghosting issues than my usb gaming keyboard.

Yes, three modifier keys at once work fine.  Matter of fact, I'm trying to use the same key combo I used in darktable v3.6.1 (Ctrl+Alt+Shift+I) for the Import function.  In the previous version, it worked great.  If it helps, I'm using a Logitech G19 keyboard, which features five-key antighosting, or whatever you call it.

When upgrading from v3.6.1 to v3.8.0, I did choose to uninstall it during the installation process, so theoretically it should be a fresh install as well.

**Comment by bengtfalke** [comment]
Defining my own shortcuts works fine on both Windows 10 and 11 here.....

**Comment by dterrahe** [comment]
Trying to find a way to reproduce to be able to search for possible causes and solutions and to be able to test if they work. Is there anything that you can think of that is specific to your system that might make it behave differently with regard to keyboard input than our systems?

Where did you install from? If self compiled(?), which version of Gtk are you using? Which language/keyboard layout settings are you using? Does switching to US English make a difference? Are you able to test with a different keyboard (not that I think the G19 would be to blame, at all, just that we could learn something by varying different aspects of the setup).

**Comment by Clorix1** [comment]
There isn't anything I can think of that would affect my keyboard behavior.  I'm running Windows 10 Pro with the US English language.  The only controller-type peripherals I have connected are a Stream Deck and a joystick and throttle for use in my flight sim.

I downloaded and installed DT through the link provided in the [reddit post](https://www.reddit.com/r/DarkTable/comments/rnkk9e/darktable_380_release/) about the update.  I downloaded the win64.exe file attached to that link's GitHub Assets section at the bottom.

I hooked up an old Dell keyboard I had lying around and it exhibits the same behavior.  Both keyboards (the G19 and Dell) are known good keyboards, and I've never had any issues with them in any software.

The only other input device I have software installed for is a Huion drawing tablet that I very rarely use.  It isn't currently connected to my PC, so I wouldn't think it would affect Darktable in any way.

It also doesn't matter if I try creating the keyboard shortcut using my actual keyboard or the Stream Deck's buttons.  I get the same effect.

**Comment by dterrahe** [comment]
That sounds eerily similar to my setup (I also have software installed for a huion that I currently don't have access to), but I'm not seeing the issues you are. I would say unfortunately, because it prevents me from diagnosing them.

Just because I'm otherwise running out of ideas; you are not running anything in the background that captures/translates keystrokes? Like keyboard macro apps? Possibly something related to the Stream Deck? Could you temporarily disable it? It would also interfere with 3.6.1 so it is unlikely to cause the different behavior, but better to exclude that possibility.

**Comment by Clorix1** [comment]
The only other software I can think of that globally "listens" for shortcuts would be ShareX and Nvidia's Shadowplay, both of which worked well alongside v3.6.1.

I just remembered during the update process, I chose to uninstall DT as part of it, and then accidentally installed the new version to the wrong drive.  So I then, without having run the new install of DT in the wrong location, once again ran the updater, choosing to uninstall DT, and then reinstalled it in the correct location.  I don't know if that might have something to do with it.

**Comment by dterrahe** [comment]
Still, if you could temporarily disable everything that processes or sends keystrokes and see if that makes any difference. There must be _something_ that is different between your setup and mine...

If apart from shortcuts dt works fine, I can't really see how your installation process would have caused this issue.

**Comment by Clorix1** [comment]
I've quit everything I had running (ShareX, Nvidia's Shadowplay overlay, Spotify, Discord and Chrome) and DT is still exhibiting the same behavior.  I'm not sure if DT will run in Safe Mode, but I'll even give that a shot and see what I can see (if anything!).

**Comment by ptilopteri** [comment]
install dt on another machine, a similar machine.

**Comment by Clorix1** [comment]
To my surprise, Darktable opened in Safe Mode, however it is STILL not correctly applying my shortcuts.

> install dt on another machine, a similar machine.

Unfortunately I don't have another computer available to try that one.  That was one of the first things I thought to do :(

I will try uninstall DT and installing it again and see what happens.

**Comment by Clorix1** [comment]
Uninstalling and freshly installing again didn't fix it either.

**Comment by Clorix1** [comment]
I've [recorded a video](https://www.youtube.com/watch?v=SWQGJ-QkOLA) just in case anyone can spot something I'm doing wrong.  In it, I have a keyboard visualizer to show exactly what keys I'm pressing on my keyboard.  The first thing I show is that the Copy & Import key is set by default (Shift+Ctrl+I) and will subsequently come up as a conflict, proving that it isn't simply a display bug.  And then you can see how even when pressing Ctrl+Shift+Alt+I results in at least two differing key combos being detectied/set in DT.

I also show how the Stream Deck software has no issues detecting all four button presses.

I hope the video helps in some way!

**Comment by dterrahe** [comment]
You're not doing anything wrong. I'm just completely mystified by this.

**Comment by Clorix1** [comment]
My apologies, but I lied.  I DO in fact have another computer that I forgot about due to it sitting in a corner and nearly never used.  Anyway, I tried it on that laptop and oddly I'm getting the same behavior on that one as well.  Totally separate PC, keyboard and all (obviously), and no peripherals connected on it.  I even tried it using Windows Sandbox and am getting the same thing everywhere.  I'm at a loss.

**Comment by dterrahe** [comment]
Hm. Maybe I figured out what's going on. Do you possibly lift the modifier keys _before_ releasing the letter key? I mean is your sequence something like:
+shift
+ctrl
+alt
+I
-alt
-I
-ctrl
-shift
?

Input NG triggers key press action on _release_ of the key (and only after a brief delay if double/triple presses mappings exist) because only then does it know if you meant a long or multi press. The modifiers are used as active at that point. This makes sense for more complicated shortcuts (key+mouse button+move) because, using fallbacks, you can for example change the speed of adjustments by adding or releasing modifier keys while you move, without having to repeat the possibly long shortcut sequence. But for key press only shortcuts, the initial modifier status could be saved and used. Could possibly introduce bugs though, so I'd have to look at that carefully.

**Comment by Clorix1** [comment]
Typically when I'm setting shortcuts in any software, I methodically hold each modifier key one-by-one and then just quickly press the final key.  So in my case, I'll start with, say, Ctrl, then Alt, then Shift, and then finally just quickly tap I before letting go of the modifier keys all at once.

However, because of your reply, I tried all sorts of methods, including letting go of each key one-by-one as well.  The shortcut doesn't get applied (detected?) until the first time releasing a modifier key.  Using your above example, this is what I get:

+shift
+ctrl
+alt
+I
-I
-Shift  (This step is where the shortcut is saved/detected/applied, and also happens to ALWAYS be the key that's missing from the shortcut!)
-alt
-ctrl

In other words, the first modifier key I release after my final key (Shift in the above example, but can be Alt or Ctrl as well) is the one that gets removed from the saved shortcut, so I definitely think you're onto something here!

I hope I've explained that clearly enough.  I'm pretty bad at explaining things.  If you need clarification I'm happy to help.

**Comment by dterrahe** [comment]
> I'm pretty bad at explaining things.

Absolutely not; it is very helpful and we keep making progress, I just haven't figured out what's happening in your case yet. Because it seems to be different than what's happening on my machine. But since both of your machines behave like this...

If you use a shortcut without modifiers (for example "f" in lighttable to switch to preview) does it immediately respond after you release the key or is there a delay? The code uses gtk-double-click-time to destinguish double/triple/long clicks. Do you have this set to a normal level (in windows mouse properties double-click speed)?

**Comment by Clorix1** [comment]
If I just quickly press F in lighttable view it switches immediately after releasing the key.  I don't notice any delay at all.  If I hold F for about a second, then it doesn't respond, which I assume means a long-press F could do something differently but is currently not set to do anything.

[Here are the settings](https://i.imgur.com/3TvrGA2.png) for my mouse click speeds and such.  I can't remember if I've edited them a long time ago, but it feels "normal" to me when double-clicking.

**Comment by Clorix1** [comment]
Just curious if you've had any breakthroughs on this issue.  I don't think I've seen anyone else mention a similar problem so I totally understand if it has gone a bit cold at the moment.  Thanks for looking into it!

**Comment by dterrahe** [comment]
No, I'm sorry, I still haven't figured out what might be causing your issue or how to reproduce it on my machine. Which means I can't play around to see if I can make it go away either.

If you were able to compile from source, I could suggest trying out one small change. But given the detailed description you have given of what's happening on your machines, I would _not_ expect it to make much difference. And setting up a compilation environment on Windows is a bit of work, to say the least. So you would likely be adding to your problems instead of solving them.

Maybe you could attract other people with the same problem who could help solve it by changing the title of this issue to "keyboard shortcuts under Windows missing modifiers".

**Comment by Clorix1** [comment]
Hey no worries.  I'd be happy to try compiling from source code if I knew how haha.  The only bit of programming I've done is a super simple program for myself using the free version of Visual Studio 2017 and I barely know how to make that work.

I'll try your suggestion of changing the title to see if it might catch anyone's eye.

Thanks again, and happy new year!

**Comment by jec993** [comment]
Howdy,

Also a Windows user (win11) running a fresh install of 3.8.1 and am having shortcuts with multiple modifiers not working as expected.

I can neither define shortcuts with more than one modifier key nor can I execute a default shortcut which has multiple modifiers.

for example, ctrl+c works as expected (can see the "copy" button even indicate the action) but ctrl+shift+c does nothing (and no indication on the "selective copy..." button)

in the shortcuts editor, shift+ctrl+c is shown to be assigned to "utility modules/history stack/copy parts" and is set to activate.

thank you.

**Comment by Clorix1** [comment]
I just installed darktable v4.0.0 and tested the keyboard shortcuts.  I'm pleased to say they appear to be working properly again!  Thanks!

**Comment by ptilopteri** [comment]
if issue is solved, please consider closing this issue.
